First implementation

benadida opened the pull request:
Basic skeleton of server and ported prototype implementation of scanner

benadida commented:
@beausmith I'd love your review on this, even though I haven't implemented test coverage yet and won't have 100% until I add more, mostly because I want to make sure the overall setup looks good to you before I go further.

beausmith commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+      }
+    })
+    
+    votes["_precinctId"] = precinctId
```
What is the `_` prefix used to convey?

benadida replied:
I was trying to set it apart from the raw data in the ballot.

benadida replied:
This format decision affects the whole chain into the EMS, too, so I'd like to keep it this way unless we have a very good reason to change it. It does make sense to me to have this metadata be different, and I was trying not to complicate the data structure and keep it flat.

beausmith commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+
+let watcher:chokidar.FSWatcher, election:Election
+
+export function init(e:Election, directoryToWatch:string, callback:BallotCallbackFunction) {
```
What do you think about passing an object as a single param if there are more than 2 params?

benadida replied:
Feels like an extra type just for the sake of ... naming the arguments? I guess, doesn't feel like a clear win.

beausmith replied:
Ok, just note that without this it requires readers of the code to remember what each param is when this function is called later. How many params are too many?

benadida replied:
in the new version, it's now reduced to just two arguments: a `path`, and a `ballotCallback`. I think that's probably simpler and reasonably future-proof rather than add an interface, export it, etc?

beausmith commented on src/scanner.ts:
```diff
@@ -0,0 +1,33 @@
+}
+
+export function getStatus() {
+  return new Promise((resolve, _reject) => {
```
Again, wondering… what is the `_` prefix used to convey?

benadida replied:
In this case it's to prevent tsc from complaining about the unused variable while I'm still implementing this function :)

beausmith commented on src/scanner.ts:
```diff
@@ -0,0 +1,33 @@
+
+export function getStatus() {
+  return new Promise((resolve, _reject) => {
+    resolve({numBallots:10})
```
Hardcoded "10"? Looks like it's already in mockScanner below.

benadida replied:
yep, hardcoded cause this was a placeholder while implementation is still happening.

beausmith commented on src/server.ts:
```diff
@@ -0,0 +1,58 @@
+
+app.use(express.json())
+
+app.get("/", (_request : Request, response : Response) => {
```
If multiple modules are being used, won't this be a shared endpoint? Is there issue of conflicting endpoints?

benadida replied:
yeah, this is just for the sake of testing the server manually in a standalone way, without going through the proxy. This makes development way easier. I actually *don't* want this page exposed to the main host.

benadida commented on README.md:
```diff
@@ -4,9 +4,32 @@ This web server component provides a web interface to a scanner
+* `POST /scan/scan` scans a batch of ballots and stores them in the scanner's database
+* `GET /scan/status` returns status information about scanned ballots so far
+* `POST /scan/export` saves all the CVRs to a USB stick
+* `POST /scan/zero` zero's all data
```
probably yes, just keep in mind this isn't the final thing, I just wanted to get a first-pass implementation in, and then iterate based on UI needs.

benadida commented on package.json:
```diff
@@ -0,0 +1,38 @@
+  },
+  "scripts": {
+    "build-ts": "tsc",
+    "watch-ts": "tsc -w",
```
sure

benadida commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+}
+
+// @ts-ignore image type not defined
+const scan = (im:Image) => {
```
I can live with that.

benadida commented:
@beausmith this was the PR I said wasn't quite ready for review, but I forgot to unassign you :)

In any case, since you've done the work of reviewing it, I'm going to do the work of addressing all of your comments and try to forge forward! This will save time and respect your review time.

benadida commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+let watcher:chokidar.FSWatcher, election:Election
+
+export function init(e:Election, directoryToWatch:string, callback:BallotCallbackFunction) {
+  election = e
```
for now yes. Eventually it will be used to interpret the ballot.

benadida commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+  }
+  
+  const secondImage = im.crop({x: width/5, width: width/5, y:height*3/4, height: height/4})
+  const secondScan = jsQR(secondImage.data, secondImage.width, secondImage.height)
```
sure

benadida commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+    // prepare the CVR
+    let votes : Ballot = {}
+    
+    const allChoicesList = allChoices.split("/")
```
yes, we're certain about the data in the QR code. It's only integers split by `/` and `.`. QR code only includes positions of candidates in the question, not write-ins.

benadida replied:
(Also, we're going to revisit the encoding of the QR code to be more compact, so this will get rewritten, but it works with current encoding.)

benadida commented:
@beausmith I've addressed all the big comments, added linting and prettier, and responded to the couple of comments where I disagree. Let me know what you think.

beausmith commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+	}
+	
+	// neither 0 nor 1, so no answer, we still record it.
+	votes[c.id] = ""
```
This is still a blocker.

beausmith commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+	votes[c.id] = ""
+      }
+      
+      if (c.type === "candidate") {
```
Thoughts on this?

beausmith commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+      return
+    }
+    
+    const qrData : string = scanResult.data as string
```
looks fixed

beausmith commented on src/types.ts:
```diff
@@ -0,0 +1,66 @@
+  readonly ballotStyles: BallotStyle[]
+  readonly parties: Parties
+  readonly precincts: Precinct[]
+  readonly districts: District[]
```
looks fixed.

beausmith commented on src/interpreter.ts:
```diff
@@ -0,0 +1,104 @@
+    let votes : Ballot = {}
+    
+    const allChoicesList = allChoices.split("/")
+    contests.forEach((c : Contest, contest_num : number) => {
```
looks fixed.

beausmith commented:
Sorry for the pre-mature review… forgot that this wasn't ready.

And, whoops… I re-requested a review of this from myself. 😛 

Please comment to let me know when this is ready for review.
